mtupper/homework.md

```diff
@@ -0,0 +1,292 @@
+### Homework #1
+
+(Due Monday 1/12)
+
+---
+
+**1)** (_Difficulty: mixed_)
+Predict the output of each of the following expressions, and explain the rules which determine it.
+If the output depends on the value of x, identify the conditions when the expression will output true.  Assume the cases are independent, and x is reset to an unknown value before each.
+
+Some of these are tricky!  Don't trust your first instinct.
+
+**a)** `"1" == 1`<br />
+> **true** - string 1 is "similar enough" to number 1<br />
+
+**b)** `"1" === 1`<br />
+> **false** - string 1 is, strictly speaking, not the same as number 1<br />
+
+**c)** `x == 'x'`<br />
+> **false** - the only way to make this true would be to declare a variable x with a value of the string 'x'<br />
+> `var x = "x"`<br />
+
+**d)** `x == (x+'')`<br />
+> **true** - since adding nothing (empty string `''`) to the variable x won't change its value<br />
+
+**e)** `'' == ' '`<br />
+> **false** - because an empty string is not the same as a string consisting of a space character<br />
+
+**f)** `x = true`<br />
+> **true** - since we just set the value of x equal to true, it is indeed true<br />
+
+**g)** `var x; x == 'undefined'`
+> **false** - while it was not assigned a value, the variable x has been defined, same as assigning value `''`<br />
+
+**h)** `'9'<'10'`<br />
+> **false** - because they are both a `"string"`, strings are ordered alphabetically (1 before 9) and therefore, 10 comes before 9<br />
+
+**i)** `typeof x + 1 === "number"`<br />
+> **false** - ```typeof x + 1``` returns "number1", adding 1 to the result of typeof x<br />
+
+**j)** `typeof x % 2 === "number"`<br />
+> **false** - ```typeof x % 2``` returns NaN because it is the same as ```"number" % 2```<br />
+
+**k)** `typeof (x % 2) === "number"` <br />
+> **true** - makes sense when x is already a number
+>
+> ```javascript
+> var x = 4;
+> typeof (x % 2); // returns "number"
+> ```
+> However, I am not clear why this `typeof` returns `"number"`
+>```javascript
+> var x = "text";
+> (x % 2); // returns NaN, but...
+> typeof (x % 2); // returns "number"
+>```
+
+**l)** `x++ == ++x`
+> **false** - value of x is returned and then incremented and incremented again before being compared to the new value of x
+>
+> ```javascript
+> var x = 1;
+> var y = x++; // y = 1, x = 2
+> var z = ++x; // z = 3, x = 3
+> y == z; // is false
+> ```
+
+**m)** `++x == x++`
+> **true** - value of x is incremented and returned, new value of x is compared to x (true), then incremented.
+>
+> ```javascript
+> var x = 1;
+> var y = ++x; // y = 2, x = 2;
+> var z = x++; // z = 2, x = 3;
+> y == z; // is true
+> ```
+
```

Comment:
Careful: x could be a non-numeric string, which doesn't increment, so ++x is NaN, and NaN != NaN.